torchqdynamics/odeint.py

```diff
@@ -157,6 +165,156 @@ def _fixed_odeint(
+    g0: Tuple[Tensor, ...],
+    t_span: Tensor,
+    dt: float,
+    parameters: Tuple[nn.Parameter, ...],
+):
+    """Integrate the augmented ODE backward using a fixed time step solver."""
+    # check t_span
+    if not (t_span.ndim == 1 and len(t_span) == 2):
+        raise ValueError(
+            f'`t_span` should be a tensor of shape (2,), but has shape {t_span.shape}.'
+        )
+    if t_span[1] <= t_span[0]:
+        raise ValueError('`t_span` should be sorted in ascending order.')
+
+    T = t_span[0] - t_span[1]
+    if not torch.allclose(torch.round(T / dt), T / dt):
+        raise ValueError('The total time of evolution should be a multiple of dt.')
+
+    # define time values
+    times = torch.linspace(t_span[0], t_span[-1], torch.round(T / dt).int() + 1)
+
+    # run the ode routine
+    y, a, g = y0, a0, g0
+    y, a = y.requires_grad_(True), a.requires_grad_(True)
+    for t in tqdm(times[:-1], leave=False):
+        with torch.enable_grad():
+            # compute y(t-dt) and a(t-dt) with the qsolver
+            y, a = qsolver.backward_augmented(t, dt, y, a, parameters)
+
+            # compute g(t-dt)
+            dg = torch.autograd.grad(
+                a, parameters, y, allow_unused=True, retain_graph=True
+            )
+            dg = none_to_zeros_like(dg, parameters)
+            g = add_tuples(g, dg)
+
+    return y, a, g
+
+
+class ODEIntAdjoint(torch.autograd.Function):
+    """Class for ODE integration with a custom adjoint method backward pass."""
+
+    @staticmethod
+    def forward(ctx, qsolver, y0, t_save, exp_ops, save_states, *parameters):
+        """Forward pass of the ODE integrator."""
+        # save into context for backward pass
+        ctx.qsolver = qsolver
+        ctx.t_save = t_save if save_states else t_save[-1]
+
+        # solve the ODE forward without storing the graph of operations
+        y_save, exp_save = _odeint_inplace(qsolver, y0, t_save, exp_ops, save_states)
+
+        # save results and model parameters
+        ctx.save_for_backward(y_save, *parameters)
+
+        return y_save, exp_save
+
+    @staticmethod
+    def backward(ctx, *grad_y):
+        """Backward pass of the ODE integrator.
+        An augmented ODE is integrated backwards starting from the final state computed
+        during the forward pass. Integration is done in multiple sequential runs
+        between every checkpoint of the forward pass, as defined by `t_save`. This
+        helps with the stability of the backward integration.
+        Throughout this function, `y` is the state, `a = dL/dy` is the adjoint state,
+        and `g = dL/dp` is the gradient w.r.t. the parameters, where `L` is the loss
+        function and `p` the parameters.
+        TODO: Check that stacking `y` and `a` does not improve performance of the
+        solver. There was funky stuff happening in this regard in some tests I ran.
+        """
+        # disable gradient computation
+        torch.set_grad_enabled(False)
```

Comment:
Should we use this as a context manager rather than just disabling it?

Reply:
Well here we deactivate the gradient calculation for the user for the rest of his session no? I don't think we should change global settings without consent.

I don't know the difference between `with torch.no_grad()` and `with torch.set_grad_enabled(True)`, you might want to read https://pytorch.org/docs/stable/notes/autograd.html#locally-disable-grad-doc to pick the correct one.